application/transaction.go

```diff
@@ -18,21 +26,33 @@ func Transactional(db DB, todo func(f Application) error) error {
 		return errors.WithStack(err)
 	}
 
-	defer func() {
-		switch err {
-		case nil:
-			log.Debug(nil, map[string]interface{}{}, "Commit the transaction!")
-			err = tx.Commit()
-		default:
-			log.Debug(nil, map[string]interface{}{}, "Rolling back the transaction...")
-			_ = tx.Rollback()
-			log.Error(nil, map[string]interface{}{
-				"err": err,
-			}, "database transaction failed!")
-			err = errors.WithStack(err)
+	return func() error {
+		errorChan := make(chan error, 1)
+		txTimeout := time.After(databaseTransactionTimeout)
+
+		go func(tx Transaction) {
+			errorChan <- todo(tx)
+		}(tx)
+
+		select {
+		case err := <-errorChan:
+			if err != nil {
+				log.Debug(nil, nil, "Rolling back the transaction...")
+				tx.Rollback()
+				log.Error(nil, map[string]interface{}{
+					"err": err,
+				}, "database transaction failed!")
+				return errors.WithStack(err)
+			}
+
+			tx.Commit()
+			log.Debug(nil, nil, "Commit the transaction!")
+			return nil
+		case <-txTimeout:
+			log.Debug(nil, nil, "Rolling back the transaction...")
+			tx.Rollback()
+			log.Error(nil, nil, "database transaction timeout!")
+			return errors.New("database transaction timeout!")
```

Comment:
@xcoulon We could generally pass the context to the transaction tand hat will solve this issue in the future. But we depend on the pq version and the one supported by gorm